### DISCUSSION TITLE
Multiple pre-installed node versions with binarySource=global

### DISCUSSION THREAD

**Discussion by DjordyKoert** [renovatebot/renovate#25228] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.81.0-full

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

We are currently running the renovate using the `renovate/renovate:36.81.0-full` image. This contains a pre-installed version of node 18.

We have a project which has a node version restriction to node `^20.0` inside of the `package.json` `engines.node` field.

When renovate attempts to do its lock-file-maintenance on the project it causes the npm command to fail because of a version mismatch.

We assumed that Renovate would detect the correct node version from the pre-installed versions for each tool when using [binarySource=global](https://docs.renovatebot.com/self-hosted-configuration/#binarysource) and use the corresponding pre-installed version of the tool.

Any idea on how we can make this work?

### What you tried so far.

We have tried to manually install node 20 inside our own docker image based on `renovate/renovate:36.81.0-full` by adding `RUN install-tool node v20.6.1`. This causes two versions to be installed inside of the container (`18.17.1` & `20.6.1`). Running `node --version` outputs: `v20.6.1`

This works fine and causes the earlier mentioned npm error to not appear but it does cause another problem because `RE2` becomes unavailable because it was installed with a different node version.
<details><summary>RE2 Logs</summary>

```
WARN: RE2 not usable, falling back to RegExp
       "err": {
         "code": "ERR_DLOPEN_FAILED",
         "message": "The module '/opt/containerbase/tools/renovate/36.81.0/node_modules/re2/build/Release/re2.node'\nwas compiled against a different Node.js version using\nNODE_MODULE_VERSION 108. This version of Node.js requires\nNODE_MODULE_VERSION 115. Please try re-compiling or re-installing\nthe module (for instance, using `npm rebuild` or `npm install`).",
         "stack": "Error: The module '/opt/containerbase/tools/renovate/36.81.0/node_modules/re2/build/Release/re2.node'\nwas compiled against a different Node.js version using\nNODE_MODULE_VERSION 108. This version of Node.js requires\nNODE_MODULE_VERSION 115. Please try re-compiling or re-installing\nthe module (for instance, using `npm rebuild` or `npm install`).\n    at Object.Module._extensions..node (node:internal/modules/cjs/loader:1327:18)\n    at Module.load (node:internal/modules/cjs/loader:1091:32)\n    at Function.Module._load (node:internal/modules/cjs/loader:938:12)\n    at Module.require (node:internal/modules/cjs/loader:1115:19)\n    at require (node:internal/modules/helpers:130:18)\n    at Object.<anonymous> (/opt/containerbase/tools/renovate/36.81.0/node_modules/re2/re2.js:3:13)\n    at Module._compile (node:internal/modules/cjs/loader:1241:14)\n    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1295:10)\n    at Module.load (node:internal/modules/cjs/loader:1091:32)\n    at Function.Module._load (node:internal/modules/cjs/loader:938:12)\n    at Module.require (node:internal/modules/cjs/loader:1115:19)\n    at require (node:internal/modules/helpers:130:18)\n    at re2 (/opt/containerbase/tools/renovate/36.81.0/node_modules/renovate/lib/expose.cjs:14:10)\n    at Object.<anonymous> (/opt/containerbase/tools/renovate/36.81.0/node_modules/renovate/lib/util/regex.ts:13:20)\n    at Module._compile (node:internal/modules/cjs/loader:1241:14)\n    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1295:10)\n    at Module.load (node:internal/modules/cjs/loader:1091:32)\n    at Function.Module._load (node:internal/modules/cjs/loader:938:12)\n    at Module.require (node:internal/modules/cjs/loader:1115:19)\n    at require (node:internal/modules/helpers:130:18)"
       }
```

</details>

We have tried rerunning `install-tool node v18.17.1` after the above mentioned install-tool command for node v20 in order to make `node` use `18.17.1` in hopes of:
1. Fixing the RE2 issue. 
2. Renovate 'finding' the node 20 requirement from the `package.json` and running npm using node 20 (located in `/opt/containerbase/tools/node/20.6.1/bin`). This is sadly not the cause.

Renovate could use the correct version of node by prepending the PATH env to let it use the correct node version when using npm. We have tried doing this manually which using `PATH="/opt/containerbase/tools/node/20.6.1/bin:$PATH" npm install --package-lock-only --no-audit --ignore-scripts` and this worked perfectly but we also understand that this might not be the ideal solution for Renovate.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: rawExec err (repository=group/renovate-test, branch=renovate/lock-file-maintenance)
       "err": {
         "cmd": "/bin/sh -c npm install --package-lock-only --no-audit --ignore-scripts",
         "stderr": "npm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: undefined\nnpm ERR! notsup Not compatible with your version of node/npm: undefined\nnpm ERR! notsup Required: {\"node\":\"^20.0\"}\nnpm ERR! notsup Actual:   {\"npm\":\"9.8.1\",\"node\":\"v18.17.1\"}\n\nnpm ERR! A complete log of this run can be found in: /tmp/renovate/cache/others/npm/_logs/2023-10-16T06_16_21_185Z-debug-0.log\n",
         "stdout": "",
         "options": {
           "cwd": "/tmp/renovate/repos/gitlab/group/renovate-test/src/frontend",
           "encoding": "utf-8",
           "env": {
             "NPM_CONFIG_CACHE": "/tmp/renovate/cache/others/npm",
             "npm_config_store": "/tmp/renovate/cache/others/pnpm",
             "HOME": "/root",
             "PATH": "/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/opt/containerbase/tools/python/3.11.5/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/opt/containerbase/tools/python/3.11.5/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/opt/containerbase/tools/python/3.11.5/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/opt/containerbase/tools/python/3.11.5/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
             "LC_ALL": "C.UTF-8",
             "LANG": "C.UTF-8"
           },
           "maxBuffer": 10485760,
           "timeout": 900000
         },
         "exitCode": 1,
         "name": "ExecError",
         "message": "Command failed: npm install --package-lock-only --no-audit --ignore-scripts\nnpm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: undefined\nnpm ERR! notsup Not compatible with your version of node/npm: undefined\nnpm ERR! notsup Required: {\"node\":\"^20.0\"}\nnpm ERR! notsup Actual:   {\"npm\":\"9.8.1\",\"node\":\"v18.17.1\"}\n\nnpm ERR! A complete log of this run can be found in: /tmp/renovate/cache/others/npm/_logs/2023-10-16T06_16_21_185Z-debug-0.log\n",
         "stack": "ExecError: Command failed: npm install --package-lock-only --no-audit --ignore-scripts\nnpm ERR! code EBADENGINE\nnpm ERR! engine Unsupported engine\nnpm ERR! engine Not compatible with your version of node/npm: undefined\nnpm ERR! notsup Not compatible with your version of node/npm: undefined\nnpm ERR! notsup Required: {\"node\":\"^20.0\"}\nnpm ERR! notsup Actual:   {\"npm\":\"9.8.1\",\"node\":\"v18.17.1\"}\n\nnpm ERR! A complete log of this run can be found in: /tmp/renovate/cache/others/npm/_logs/2023-10-16T06_16_21_185Z-debug-0.log\n\n    at ChildProcess.<anonymous> (/opt/containerbase/tools/renovate/36.81.0/node_modules/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:526:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)"
       },
       "durationMs": 198412
```

</details>

**Comment by rarkins** [comment]
When talking of the RE2 problem, do you mean that Renovate itself starts up using the node v20 version? If so then maybe you can change Renovate's startup command so that it uses the node v18 path to run instead of default `node`

**Reply by DjordyKoert** [reply]
Running `PATH="/opt/containerbase/tools/node/18.17.1/bin:$PATH" renovate` does fix the RE2 issue but still causes NPM to fail while updating the project because the project requires node `^20.0` to be used while npm is running on `18.17.1`.

Normally when using `binarySource=install` [the required tool would be installed during runtime](https://github.com/renovatebot/renovate/blob/270e2964466c147effde28f23badd5abe529ee80/lib/util/exec/index.ts#L76). But we are using `binarySource=global` with all required tool versions already installed. We expected renovate to do some check or something to see if the required version is available (and if it is not available fallback to the default version of the tool) and use the correct tool version but this seems to not be the case by following the code

**Reply by rarkins** [reply, marked as answer]
Your "expectations" are unfortunately too optimistic, feel free to investigate the code and submit a PR if you can find a way to get it to work like you expect.

**Reply by DjordyKoert** [reply]
My bad 'expected' wasn't the correct word for it we more hoped that it would work that way. Will see if I can find a way to make it work
